File: Problems/LinkedList/CloneListWithRandomPreviousPointer.py
import sys
class node:

    def __init__(self, info):
        self.info = info
        self.next = None
        self.random=None


class LinkedList:

    def __init__(self):
        self.head = None


    def display(self):
        temp = self.head
        while (temp):
            print(temp.info)
            temp = temp.next

    def count(self):
        count=0
        temp=self.head
        while(temp):
            count+=1
            temp=temp.next
        return count

    def serach(self,value):
        temp=self.head
        pos=1
        while(temp):
            if(temp.info==value):
                print("The value found at",pos)
            temp=temp.next
            pos+=1
        print("Value not found in the linked list")
    def insert_at_beg(self,data):
        self.temp = node(data)
        if self.head is None:
            self.head = self.temp
            return
        self.temp.next = self.head
        self.head= self.temp
    def insert_at_end(self,data):
        self.temp = node(data)
        if self.head is None:
            self.head = self.temp
            return
        self.p = self.head
        while self.p.next is not None:
            self.p=self.p.next
        self.p.next = self.temp;

    def insert_after_given_node(self,data,item):
        self.p=self.head
        while self.p is not None:
            if(self.p.info==item):
                self.temp=node(data)
                self.temp.next=self.p.next
                self.p.next=self.temp
                return
            self.p=self.p.next
        print("Item not found")

    def insert_at_pos(self,data,pos):
        self.temp=node(data)
        if pos==1:
            self.temp.next = self.head
            self.head= self.temp
            return
        self.p=self.head
        while pos>2 and self.p is not None:
            self.p=self.p.next
            pos-=1
        if self.p is None:
            print("Position exceeded the length of the list")
        else:
            self.temp.next=self.p.next
            self.p.next=self.temp
    def delete(self,data):
        if self.head is None:
            print("List is empty")
            return
        if self.head.info==data:
            self.temp=self.head
            self.head=self.head.next
            return
        self.p=None
        self.curr=self.head
        while self.curr is not None:
            if self.curr.info==data:
                self.temp=self.p.next
                self.p.next=self.temp.next
                return
            self.p=self.curr
            self.curr=self.curr.next
def clone(head):
    curr = head
    while curr != None:
        new = node(curr.info)
        new.next = curr.next
        curr.next = new
        curr = curr.next.next
    curr = head
    while curr != None:
        curr.next.random = curr.random.next
        curr = curr.next.next

    curr = head
    dup_root = head.next
    while curr.next != None:
        tmp = curr.next
        curr.next = curr.next.next
        curr = tmp

    return dup_root
def print_list(head):
    temp=head
    while temp!=None:
        print('Data =', temp.info, ', Random =', temp.random.info)
        temp=temp.next

if __name__=='__main__':
    llist=LinkedList()
    for i in range(5):
        llist.insert_at_end(int(input()))
    temp=llist.head
    temp.random = temp.next.next
    temp.next.random = temp
    temp.next.next.random = temp.next.next.next.next
    temp.next.next.next.random = temp.next.next.next.next
    temp.next.next.next.next.random = temp.next
    print("original list")
    print_list(llist.head)
    cloned_list = clone(llist.head)

    print('\nCloned list:')
    print_list(cloned_list)

File: Problems/LinkedList/CloneListWithRandomPreviousPointer.go
package main

import (
    "errors"
    "fmt"
)


type Node struct{
    Data int
    Next  *Node
    Prev  *Node
}

func New(i int) *Node{
    n:=Node{Data: i,Next: nil,Prev: nil}
    return &n
}



type DoublyLinkedList struct {
    head  *Node
}

func (l *DoublyLinkedList) InsertAtBeginning(i int)error{
    if l==nil{
        return errors.New(" LinkedList Not initialzed")
    }
    data:=New(i)
    if l.head!=nil{
        temp:=l.head
        l.head=data
        data.Next=temp
        temp.Prev=data
    }else{
        l.head=data
    }
    return nil
}

func (l *DoublyLinkedList) InsertAtEnd(i int)error{
    if l==nil{
        return errors.New(" LinkedList Not initialzed")
    }
    data:=New(i)
    if l.head!=nil{
        temp:=l.head
        for temp.Next!=nil{
            temp=temp.Next
        }
        temp.Next=data
    }else{
        l.head=data
    }
    return nil
}

func (l *DoublyLinkedList) DeleteNode(i int)error{
    if l==nil{
        return errors.New(" LinkedList Not initialzed")
    }
    if l.head==nil{
        return errors.New(" LinkedList Empty")
    }
    if l.head.Next==nil && l.head.Data==i{
        fmt.Println("Deleting Data from Head ",l.head.Data)
        l.head=nil
        return nil
    }
    data:=l.head
    for data!=nil{
        if data.Data==i{
            data.Prev.Next=data.Next
            data.Next.Prev=data.Prev

            return nil
        }
        data=data.Next

    }
    if data.Data==i{
        data.Prev.Next=nil
    }
    fmt.Println("Element not found in list ")
    return nil
}

func (l *DoublyLinkedList) Search(i int)error{
    if l==nil{
        return errors.New(" LinkedList Not initialzed")
    }
    if l.head==nil{
        return errors.New(" LinkedList Empty")
    }
    if l.head.Data==i{
        fmt.Println("Element Found ")
        return nil
    }
    data:=l.head
    for data.Next!=nil{
        if data.Next.Data==i{
            fmt.Println("Element Found ")
            return nil
        }
        data=data.Next
    }
    fmt.Println("Element not found in list ")
    return nil
}


func (l *DoublyLinkedList) display()error{
    if l==nil{
        return errors.New(" LinkedList Not initialzed")
    }

    if l.head!=nil{
        data:=l.head
        for data!=nil{
            fmt.Println(data.Data)
            data=data.Next
        }
    }else{
        errors.New(" LinkedList Empty")
    }
    return nil
}

func cloneLinkedList(l *DoublyLinkedList) *DoublyLinkedList{
    tmp:=l.head
    for tmp!=nil{
        newNode:=New(tmp.Data)
        newNode.Next=tmp.Next
        tmp.Next=newNode
        tmp=tmp.Next.Next
    }
    tmp=l.head
    for tmp!=nil{
      tmp.Next.Prev=tmp.Prev.Next
      tmp=tmp.Next.Next
    }

    tmp=l.head
    dup_root := tmp.Next
    for tmp.Next!=nil{
        curr:=tmp.Next
        curr.Next=curr.Next.Next
        curr=tmp
    }
    return &DoublyLinkedList{dup_root}
}

func main(){

    var l DoublyLinkedList
    l.InsertAtEnd(12)
    l.InsertAtEnd(14)
    l.InsertAtEnd(16)
    l.InsertAtEnd(18)
    l.InsertAtEnd(20)
    l.InsertAtEnd(40)

    l.head.Prev = l.head.Next.Next
    fmt.Println()
    l.head.Next.Prev = l.head
    fmt.Printf(" %d --->%d", l.head.Next.Prev.Data , l.head.Data)
    fmt.Println()
    l.head.Next.Next.Next = l.head.Next.Next.Next.Next
    fmt.Printf(" %d --->%d", l.head.Next.Next.Next.Data , l.head.Next.Next.Next.Next.Data)
    fmt.Println()
    l.head.Next.Next.Next.Prev = l.head.Next.Next.Next.Next
    fmt.Printf(" %d --->%d", l.head.Next.Next.Next.Prev.Data , l.head.Next.Next.Next.Next.Data)
    fmt.Println()
    l.head.Next.Next.Next.Next.Prev = l.head.Next
    fmt.Printf(" %d --->%d", l.head.Next.Next.Next.Next.Prev.Data, l.head.Next.Data)
    fmt.Println()
    fmt.Println("original list")
    l.display()
    fmt.Println("=====================================")
    fmt.Println("=====================================")
    copyList:=cloneLinkedList(&l)
    copyList.display()


}
